lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,446 @@
+        })
+    }
+
+    fn deleted_file_path(base: &Path) -> PathBuf {
+        base.join(DELETED_FILE_NAME)
+    }
+
+    fn version_mapping_file_path(base: &Path) -> PathBuf {
+        base.join(VERSION_MAPPING_FILE_NAME)
+    }
+
+    fn mappings_file_path(base: &Path) -> PathBuf {
+        base.join(MAPPINGS_FILE_NAME)
+    }
+
+    fn serialize_file<T: Serialize>(path: &Path, value: &T) -> OperationResult<()> {
+        let file = File::create(path)?;
+        let writer = BufWriter::new(file);
+        bincode::serialize_into(writer, value).map_err(|err| {
+            OperationError::InconsistentStorage {
+                description: format!("id tracker can't be serialized: {:?}", err),
+            }
+        })?;
+        Ok(())
+    }
+}
+
+/// Returns the required mmap filesize for a `BitSlice`.
+fn bitmap_mmap_size(deleted: &BitSlice) -> usize {
+    let num_bytes = deleted.len().div_ceil(8); // used bytes
+    num_bytes.div_ceil(8) * 8 // get next factor of 8
+}
+
+impl IdTracker for ImmutableIdTracker {
+    fn internal_version(&self, internal_id: PointOffsetType) -> Option<SeqNumberType> {
+        self.internal_to_version.get(internal_id as usize).copied()
+    }
+
+    fn set_internal_version(
+        &mut self,
+        internal_id: PointOffsetType,
+        version: SeqNumberType,
+    ) -> OperationResult<()> {
+        if self.external_id(internal_id).is_some() {
+            if let Some(old_version) = self.internal_to_version.get_mut(internal_id as usize) {
+                *old_version = version;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn internal_id(&self, external_id: PointIdType) -> Option<PointOffsetType> {
+        match external_id {
+            PointIdType::NumId(num) => self.mappings.external_to_internal_num.get(&num).copied(),
+            PointIdType::Uuid(uuid) => self.mappings.external_to_internal_uuid.get(&uuid).copied(),
+        }
+    }
+
+    fn external_id(&self, internal_id: PointOffsetType) -> Option<PointIdType> {
+        if *self.deleted.get(internal_id as usize)? {
+            return None;
+        }
+
+        self.mappings
+            .internal_to_external
+            .get(internal_id as usize)
+            .map(|i| i.into())
+    }
+
+    fn set_link(
+        &mut self,
+        _external_id: PointIdType,
+        _internal_id: PointOffsetType,
+    ) -> OperationResult<()> {
+        panic!("Trying to call a mutating function (`set_link`) of an immutable id tracker");
+    }
+
+    fn drop(&mut self, external_id: PointIdType) -> OperationResult<()> {
+        let internal_id = match external_id {
+            PointIdType::NumId(num) => self.mappings.external_to_internal_num.get(&num).copied(),
+            PointIdType::Uuid(uuid) => self.mappings.external_to_internal_uuid.get(&uuid).copied(),
+        };
+
+        if let Some(internal_id) = internal_id {
+            self.deleted.set(internal_id as usize, true);
```

Comment:
Similar to simple ID tracker, we need to ensure, that data on disk is not changed before flush. We would need to use something similar to `DatabaseColumnScheduledUpdateWrapper`

Reply:
Same goes to version mmap